tidy3d/components/source.py

```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
         return const * offset * oscillation * amp
 
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence.
+
+    Note
+    ----
+    To interpolate to simulation time steps, the value provided at the nearest time is used.
+
+    Example
+    -------
+    >>> sourcetime = CustomSourceTime.from_numpy(values=np.linspace(0, 1, 100), dt=.1)
+    """
+
+    sourcetime_dataset: Optional[TimeDataset] = pydantic.Field(
+        ..., title="Source time dataset", description="Dataset for storing the custom source time."
+    )
+
+    _sourcetime_dataset_none_warning = warn_if_dataset_none("sourcetime_dataset")
```

Comment:
do we want to add any validation on the supplied sourcetime_dataset? For example, what if the time coordinates are all negative? or all extend outside of the simulation run_time? will it be handled as expected?